### DISCUSSION TITLE
inlabru implementation

### DISCUSSION THREAD

**Discussion by karinalilleborge** [davidbolin/MetricGraph#9] [Q&A] [answered]
Hello,

I have tried to redo the inlabru implementation as demonstrated in the vignette [here](https://davidbolin.github.io/MetricGraph/articles/inlabru_interface.html#inlabru-implementation), but with some minor changes motivated by how I want to use the package. I do encounter some strange results, which makes me question if I have misunderstood some of the code and/or the way the package works - or if this is to be expected behaviour for this use?

The initial graph is the same as in the example. I then make a mesh on the graph, with `h=0.005`. From this I construct a new graph using the mesh nodes and edges:
```
# new graph where all mesh nodes are vertices
graph = metric_graph$new(V=graph0$mesh$V, 
                                               E=graph0$mesh$E, 
                                               tolerance = list(vertex_vertex=0,
                                                                         edge_edge=0,
                                                                         vertex_edge=0))
```
The random observation locations, I construct as follows (I know the (mesh)vertices I have observed on in my application, and I use the mesh when deciding the location for a given observation)
```
#simulate observations
n_obs <- 100
loc_obs <- sample(1:graph$nV, n_obs)
```
and I simulate the data as in the vignette, but using the `graph`-object from above, and I specify `PtE = VtE_mesh` which I have defined as `VtE_mesh=graph$VtEFirst()`:
```
#simulate true field for all vertices
u <- sample_spde(range = range, sigma = sigma, alpha = alpha,
                 graph = graph, PtE = VtE_mesh)
```
I would think I could get observations from the field using `loc_obs`, as this used the indices for the vertices. 
```
y <- u[loc_obs] + rnorm(n_obs, sd=sigma.e )
```
while the normalized observation locations are retrieved from `VtE_mesh[obs_loc,]`, and I use these when I add the observations to the graph before making the model and fit with inlabru.

It seems like the SPDE parameters are reasonably estimated, while the noise is always off (underestimated). 
I have attached the code as well here:
[inlabru-implementation.md](https://github.com/user-attachments/files/16027213/inlabru-implementation.md)
Specifically, I am concerned with the last plot created in the end of the given code.

I hope the question is clear, but if not I am happy to answer any questions regarding the approach.

Best regards,
Karina Lilleborge

PhD candidate,
Department of Mathematical Sciences,
Norwegian University of Science and Technology

**Comment by AlexandreBSimas** [comment, marked as answer]
Hi Karina,

I believe your problem is that the sample size is too small (we also have this problem in the standard Euclidean case). 

You can try using all the observations from the mesh you chose to see. 

For instance, with all observations (915), I got:

Precision for the Gaussian observations  4.40 0.692       3.20     4.34       5.92  4.22
(mean is 4.40)

So,
r$> sqrt(1/4.4)
[1] 0.4767313

Here with n_obs=500:
(mean of the measurement error): 4.13
r$> sqrt(1/4.13)
[1] 0.4920678

Which is really good. 

I believe this should solve your problem. Please, let me know if you find any other issue. 

Another thing that might be helpful in the future, is that the other common problem we also find (besides small sample size) is the scale between the standard deviation from the measurement error and the field itself. The standard deviation from the measurement error must always be much smaller than the one from the field, otherwise they might get confounded.
